rust/opendp/src/interactive.rs

```diff
@@ -0,0 +1,274 @@
+    }
+}
+
+type InteractiveMeasurement<DI, DO, MI, MO, S, Q> = Measurement<DI, AllDomain<Queryable<S, Q, <DO as Domain>::Carrier>>, MI, MO>;
+
+/// The state of an adaptive composition Queryable.
+pub struct AcState<DI: Domain, DO: Domain, MI: Metric, MO: Measure> {
+    input_domain: DI,
+    output_domain: DO,
+    input_metric: MI,
+    output_measure: MO,
+    d_in_budget: MI::Distance,
+    d_out_budget: MO::Distance,
+    data: DI::Carrier,
+}
+impl<DI: Domain, DO: Domain, MI: Metric, MO: Measure> AcState<DI, DO, MI, MO> where MO::Distance: MeasureDistance {
+    pub fn new(
+        input_domain: DI,
+        output_domain: DO,
+        input_metric: MI,
+        output_measure: MO,
+        data: DI::Carrier,
+        d_in_budget: MI::Distance,
+        d_out_budget: MO::Distance,
+    ) -> Self {
+        Self {
+            input_domain,
+            output_domain,
+            input_metric,
+            output_measure,
+            data,
+            d_in_budget,
+            d_out_budget,
+        }
+    }
+
+    /// Checks that a measurement (of a query) is compatible with this Queryable state.
+    fn check_types(&self, measurement: &Measurement<DI, DO, MI, MO>) -> Fallible<()> {
+        if measurement.input_domain.as_ref() != &self.input_domain {
+            return fallible!(DomainMismatch, "wrong query input domain")
+        } else if measurement.output_domain.as_ref() != &self.output_domain {
+            return fallible!(DomainMismatch, "wrong query output domain")
+        } else if measurement.input_metric.as_ref() != &self.input_metric {
+            return fallible!(DomainMismatch, "wrong query input metric")
+        } else if measurement.output_measure.as_ref() != &self.output_measure {
+            return fallible!(DomainMismatch, "wrong query output metric")
+        }
+        Ok(())
+    }
+
+    /// Checks that there is adequate budget in this Queryable state.
+    fn check_budget(&self, privacy_relation: &PrivacyRelation<MI, MO>, d_out_query: &MO::Distance) -> Fallible<()> {
+        privacy_relation.eval(&self.d_in_budget, d_out_query)?;
+        if d_out_query > &self.d_out_budget {
+            return fallible!(FailedRelation, "not enough budget")
+        }
+        Ok(())
+    }
+
+    /// Updates this Queryable state by consuming the given amount of budget.
+    fn update(self, d_out_query: &MO::Distance) -> Fallible<Self> {
+        Ok(Self { d_out_budget: self.d_out_budget.sub(d_out_query)?, ..self })
+    }
+
+    /// Processes a query, generating a new Queryable state.
+    fn transition(self, (measurement, d_out_query): &AcQuery<DI, DO, MI, MO>) -> Fallible<(Self, DO::Carrier)>
+        where MO::Distance: Clone + MeasureDistance {
+        self.check_types(measurement)?;
+        self.check_budget(&measurement.privacy_relation, d_out_query)?;
+        let res = measurement.function.eval(&self.data)?;
+        let new = self.update(d_out_query)?;
+        Ok((new, res))
```

Comment:
Great- function only runs if query is dp and budget sufficient. Budget only consumed if function evaluates.